20220714-lookup-api.md

```diff
@@ -0,0 +1,462 @@
+
+The `Lookup` endpoint (unary gRPC endpoint) will be mapped to it's HTTP/json equivalent through the mapping provided by the implementation of the [grpc-gateway](https://grpc-ecosystem.github.io/grpc-gateway/) project.
+
+## API Changes
+The following is the proposed patch diff of the protobuf API changes to [openfga/api](https://github.com/openfga/api):
+```
+diff --git a/openfga/v1/openfga_service.proto b/openfga/v1/openfga_service.proto
+index 4140d56..d0056cb 100644
+--- a/openfga/v1/openfga_service.proto
++++ b/openfga/v1/openfga_service.proto
+@@ -661,6 +661,48 @@ service OpenFGAService {
+       description: "Returns a paginated list of OpenFGA stores."
+     };
+   }
++
++  // StreamedLookup is a streaming variation of the Lookup API (see Lookup for more info).
++  rpc StreamedLookup(LookupRequest) returns (stream StreamedLookupResponse) {
++    option (google.api.http) = {
++      post: "/stores/{store_id}/lookup:stream",
++      body: "*"
++    };
++  }
++
++  // Lookup looks up the object ids for objects of the provided type that the given user
++  // has a specific relation with.
++  rpc Lookup(LookupRequest) returns (LookupResponse) {
++    option (google.api.http) = {
++      post: "/stores/{store_id}/lookup",
++      body: "*"
++    }
++  }
++}
++
++message LookupRequest {
++    string store_id  = 1 [
++      json_name = "store_id"
++    ];
++
++    string authorization_model_id = 2 [
++      json_name = "authorization_model_id"
++    ];
++
++    string object_type = 3 [
++      json_name = "object_type"
++    ];
++
++    string relation = 4;
++    string user = 5;
++
++    openfga.v1.ContextualTupleKeys contextual_tuples = 6 [
++      json_name = "contextual_tuples"
++    ];
++}
++
++message LookupResponse {
++  repeated string object_ids = 1;
++}
++
++message StreamedLookupResponse {
++  string object_id = 1;
++}
+```
+
+`store_id`, `object_type`, `relation`, and `user` will be required fields. If the `authorization_model_id` is not provided it will default to the latest model.
+
+The streaming variant accepts the same request body but differs in its response structure. For the streaming RPC a single `StreamedLookupResponse` is streamed back per result, whereas with the unary variant an array of object ids is accumulated and then returned.
+
+The `/stores/{store_id}/lookup:streamed` HTTP path mapping for the streaming variant was inspired by [Google's HTTP mapping for Custom Methods](https://cloud.google.com/apis/design/custom_methods), but this is open for debate.
+
+If anyone has any opinions or recommendations we'd love to hear them!
+
+## Configuration Changes
+- Add a `StreamedLookupDeadline` config param (`--streamed-lookup-deadline` flag)
```

Comment:
Both of these flags could be applied to their counter parts.

In a streamed scenario you may not want them all. Ultimately the client can just close the connection instead of the server doing it for them.

In the unary scenario, the user might want to specify a deadline and just use the results gathered up to that point. Closing the connection in this scenario leaves them with nothing.

Reply:
They could be applied to their counter parts, but the intent behind the two different flags is to control the semantics of the API endpoint a little differently.

The `--lookup-max-results` flag is inteded to limit the maximum number of results that can accumulate (regardless of time). This is to protect the server. If you do this with a timeout deadline instead, then there isn't necessarily a bounds for how many may accumulate in the timeout window, and that could hurt the server significantly. For example, if you accumulate 1 million lookup results in the timeout deadline then that's going to have a large memory footprint on the server.

The streaming variant with a `--streamed-lookup-deadline` flag doesn't suffer this same caveat because it doesn't have to accumulate and then respond, it just responds as each result is available. The `--lookup-max-results` flag could be used for the streaming variant and it just returns after N number of results have been streamed back. That would be a reasonable approach as well, but the semantic intent behind the deadline based streaming approach is that you may be able to stream more than `--lookup-max-results` in the deadline window, and so it may be more limiting than it needs to be 🤷 .

Those were some of the reasons behind the configuration changes that are outlined here. WDYT?

Reply:
It was already clear the intentions of each flag described.

I explained how the same could be used in their counterpart to have additional level of control. Perhaps a bit too much control for a v0.

Reply:
Correct. 

Or having them independently configurable which might be a bit too much: 
```
--lookup-max-results
--lookup-deadline
--streamed-lookup-deadline
--streamed-max-results
```

Reply:
Thanks for the clarification 👍 

I think it would be best to have a single config that controls the behavior of both for now, because this will make the two endpoints behavior more similarly and incurs less config overhead. If we find a good reason to control them independently of one another, then we can introduce the other config in the future.

Reply:
I updated this section based on this conversation. Please let me know what you think with the updates.